src/daft-plan/src/optimization/optimizer.rs

```diff
@@ -1,128 +1,598 @@
+    Once,
+    // Apply the batch of rules multiple times, to a fixed-point or until the max
+    // passes is hit.
+    // If parametrized by Some(n), the batch of rules will be run a maximum
+    // of n passes; if None, the number of passes is capped by the default max
+    // passes given in the OptimizerConfig.
+    #[allow(dead_code)]
+    FixedPoint(Option<usize>),
+}
+
 /// Logical rule-based optimizer.
 pub struct Optimizer {
-    pub rules: Vec<Arc<dyn OptimizerRule>>,
+    // Batches of rules for the optimizer to apply.
+    pub rule_batches: Vec<RuleBatch>,
+    // Config for optimizer.
+    config: OptimizerConfig,
 }
 
 impl Optimizer {
-    pub fn new() -> Self {
-        let rules: Vec<Arc<dyn OptimizerRule>> = vec![Arc::new(PushDownFilter::new())];
-        Self::with_rules(rules)
+    pub fn new(config: OptimizerConfig) -> Self {
+        // Default rule batches.
+        let rule_batches: Vec<RuleBatch> = vec![RuleBatch::new(
+            vec![Box::new(PushDownFilter::new())],
+            RuleExecutionStrategy::Once,
+        )];
+        Self::with_rule_batches(rule_batches, config)
+    }
+
+    pub fn with_rule_batches(rule_batches: Vec<RuleBatch>, config: OptimizerConfig) -> Self {
+        Self {
+            rule_batches,
+            config,
+        }
     }
-    pub fn with_rules(rules: Vec<Arc<dyn OptimizerRule>>) -> Self {
-        Self { rules }
+
+    /// Optimize the provided plan with this optimizer's set of rule batches.
+    pub fn optimize<F>(
+        &self,
+        plan: Arc<LogicalPlan>,
+        mut observer: F,
+    ) -> DaftResult<Arc<LogicalPlan>>
+    where
+        F: FnMut(&LogicalPlan, &RuleBatch, usize, bool, bool),
+    {
+        let mut plan_tracker = LogicalPlanTracker::new(self.config.default_max_optimizer_passes);
+        plan_tracker.add_plan(plan.as_ref());
+        // Fold over rule batches, applying each rule batch to the tree sequentially.
+        self.rule_batches.iter().try_fold(plan, |plan, batch| {
+            self.optimize_with_rule_batch(batch, plan, &mut observer, &mut plan_tracker)
+        })
     }
 
-    pub fn optimize(
+    // Optimize the provied plan with the provied rule batch.
+    pub fn optimize_with_rule_batch<F>(
         &self,
+        batch: &RuleBatch,
         plan: Arc<LogicalPlan>,
-        config: &OptimizerConfig,
-    ) -> DaftResult<Arc<LogicalPlan>> {
-        let mut new_plan = plan.clone();
-        let mut i = 0;
-        while i < config.max_optimizer_passes {
-            for rule in &self.rules {
-                let result = self.optimize_with_rule(rule, &new_plan);
-                match result {
-                    Ok(Some(plan)) => {
-                        new_plan = plan;
+        observer: &mut F,
+        plan_tracker: &mut LogicalPlanTracker,
+    ) -> DaftResult<Arc<LogicalPlan>>
+    where
+        F: FnMut(&LogicalPlan, &RuleBatch, usize, bool, bool),
+    {
+        let result = (0..batch.max_passes(&self.config)).try_fold(
```

Comment:
I'm certainly cool with that, I didn't add any new logging/error handling in this PR since I wanted to wait until the snafu stuff lands, but I can add logging at all of the appropriate spots in this PR if you would like.